### PR TITLE
Add Data Collection

### PR DESCRIPTION
**Pull request by emanguy**
* Add data collection object and tests
* Add AMPParse.buildUndeclaredType() function to allow creation of other model builders even if the `<script>` tags are read out of order. For instance, if dataCollection.js is included in the HTML before timeBasedRule.js it can still use `AMPParse.buildTimeBasedRule()`.

### CODE REVIEW

File: src/js/AMPSetup/AMPSetup.js [hunk 1]
```diff
@@ -78,3 +81,65 @@ AMPHexConsumer.prototype.consumeNibbles = function(nibbles)
...
+{
+	if ( !(hexadecimal instanceof AMPHexConsumer) )
+	{
+		throw new TypeError("Did not receive an instance of AMPHexConsumer");
+	}
+	if (isNaN(typeEnumeration))
+	{
+		throw new TypeError("typeEnumeration is not a number");
+	}
+
```
Comment (gti28) [inline src/js/AMPSetup/AMPSetup.js:104]: we could add another check here to verify the typeEnumeration is in the correct range

**Review by gti28 — APPROVED** [review]
looks good

File: src/js/AMPSetup/AMPSetup.js (hunk 1 above)
Comment (emanguy) [reply src/js/AMPSetup/AMPSetup.js:104]: That gets handled in the try/catch below. If we try to access a value not in the AMPBuilders object it'll throw a reference error. It changes the error message when it gets thrown out of this function.

Comment (gti28) [reply src/js/AMPSetup/AMPSetup.js:104]: makes sesne, sorry, completely missed that

**Pull request merged**